### PR TITLE
Add sensor management tutorials

### PR DESCRIPTION
**Pull request by nperree-dstl**
The first sensor management tutorial which introduces sensor management and includes an example of sensor management methods using a single sensor.

### CODE REVIEW

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+#
+# Sensor management as a POMDP
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Sensor management problems can be considered as Partially Observable Markov Decision Processes (POMDPs) where
+# observations provide information about the current state of the system but there is uncertainty in the estimate
+# of the underlying state due to noisy sensors and imprecise models of target evaluation.
+# 
+# POMDPs consist of:
+#  * :math:`X_k`, the finite set of possible states for each stage index :math:`k`.
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:50]: `stage` or `state`?

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+#  * :math:`A_k`, the finite set of possible actions for each stage index :math:`k`.
+#  * :math:`R_k(x, a)`, the reward function.
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:52]: :math:`x \in X_k` and :math:`a \in A_k`?

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+# """""""""""""
+# 
+# The first method :class:`RandomManager`, chooses a target to observe randomly. To do this the :meth:`choose_actions`
+# function uses :meth:`np.random.uniform()` to draw random samples from a uniform distribution between 0 and 1, and
+# multiply by the number of targets in the simulation. This means a single target to observe is selected randomly at
+# each time step.
+
+
+class RandomManager(Base):
+    action_list = Property(list, doc="List of possible actions")
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:256]:
```suggestion
    action_list: list = Property(doc="List of possible actions")
```

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+
+    def choose_actions(self):
+        return self.action_list[int(np.floor(len(self.action_list) * np.random.uniform()))]
+
+# %%
+# UncertaintyManager
+# """"""""""""""""""
+# 
+# The second method :class:`UncertaintyManager` selects a chosen target to observe based on the difference between the
+# covariance matrices of the prediction an the update of predicted measurement. This means the sensor manager chooses
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:266]:
```suggestion
# covariance matrices of the prediction and an update using the predicted measurement. This means the sensor manager chooses
```
?

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 5]
```diff
@@ -0,0 +1,675 @@
...
+# to observe the target for which the total uncertainty will be reduced the most by making that observation.
+# 
+# The :meth:`calculate_reward` function calculates the difference between the covariance matrix norms of the
+# prediction and the posterior assuming a predicted measurement corresponding to that prediction. The
+# :meth:`reward_list` function then generates a list of this metric for every track.
+# 
+# The :meth:`choose_actions` function takes in the variable `action_list_metric` generated by the :meth:`reward_list`
+# function and chooses the target with the largest value of this metric to observe.
+# 
+# The number of the target which is to be observed, :math:`N` is found using the following equation:
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:276]:
```suggestion
# The ID of the target which is to be observed, :math:`N` is found using the following equation:
```

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+
+from stonesoup.metricgenerator.tracktotruthmetrics import SIAPMetrics
+
+siap_generator = SIAPMetrics(position_mapping=[0, 2], velocity_mapping=[1, 3])
+
+# %%
+# The SIAP metrics require an associator to associate tracks to ground truths. This is done using the
+# :class:`~.TrackIDbased` associator. This associator uses the track ID to associate each track to the ground truth
+# with the same ID. The associator is initiated and later used in the metric manager.
+
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/01_SingleSensorManagement.py:509]: Can probably remove these metric descriptions since we have a metrics example.

**Review by oharrald-Dstl — APPROVED** [review]
Only minor comments on details. Looks good.

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py (hunk 1 above)
Comment (nperree-dstl) [reply docs/tutorials/sensormanagement/01_SingleSensorManagement.py:50]: It is referred to as "stage index" in the Foundations and Applications of Sensor Management book [1] but would time step make more sense?

Comment (oharrald-Dstl) [reply docs/tutorials/sensormanagement/01_SingleSensorManagement.py:50]: Shows my lack of knowledge! Got it.

File: docs/tutorials/sensormanagement/01_SingleSensorManagement.py (hunk 2 above)
Comment (nperree-dstl) [reply docs/tutorials/sensormanagement/01_SingleSensorManagement.py:52]: Yes - should I make that clearer in the list?

Comment (oharrald-Dstl) [reply docs/tutorials/sensormanagement/01_SingleSensorManagement.py:52]: Its probably quite obvious to most. Maybe worth it for completionist sake.

**Review by sdhiscocks — COMMENTED** [review]
First one looks good.

File: docs/tutorials/sensormanagement/02_MultipleSensorManagement.py [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+# %%
+# UncertaintyManager class
+# """"""""""""""""""""""""
+#
+# The second method :class:`UncertaintyManager` chooses the target observation configuration which results
+# in the largest difference between the uncertainty covariances of the target predictions and posteriors
+# assuming a predicted measurement corresponding to that prediction. This means the sensor manager chooses
+# to observe the targets such that the uncertainty will be reduced the most by making that observation.
+#
+# For each target in an observation configuration the :meth:`calculate_reward function` calculates the
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/02_MultipleSensorManagement.py:210]:
```suggestion
# For each target in an observation configuration the :meth:`calculate_reward` function calculates the
```

File: docs/tutorials/sensormanagement/02_MultipleSensorManagement.py [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+            metric_list.append(metric)
+        return metric_list
+
+    def calculate_reward(self, tracks_list, config, metric_time):
+        self.sensor_list = list(sensor_set)
+        config_metric = 0
+
+        # For each observation in the configuration
+        for obs in set(config):
+            sensor = self.sensor_list[config.index(obs)]
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/02_MultipleSensorManagement.py:272]:
```suggestion
        for obs, sensor in zip(list(config), self.sensor_list):
```

File: docs/tutorials/sensormanagement/02_MultipleSensorManagement.py [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+
+# %%
+# Run the sensor managers
+# ^^^^^^^^^^^^^^^^^^^^^^^
+#
+# From here the method differs slightly for each sensor manager. :class:`RandomManager` does not require any other input
+# variables whereas :class:`UncertaintyManager` requires a tracks list and a time step.
+#
+# For both sensor management methods, at each time step a prediction is made for each of the targets except the chosen
+# target,  which is updated. These states are appended to the tracks list.
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/02_MultipleSensorManagement.py:322]:
```suggestion
# For both sensor management methods, at each time step a prediction is made for each of the targets except the chosen
# targets,  which is updated. These states are appended to the tracks list.
```

File: docs/tutorials/sensormanagement/02_MultipleSensorManagement.py [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+        chosen_target = randommanager.choose_actions()
+        chosen_targets.append(chosen_target)
+
+        # The ground truth will therefore be:
+        selected_truth = truths[chosen_target][timestep]
+
+        # Observe this
+        measurement = sensor.measure([selected_truth], noise=True)
+        detection = list(measurement)
+        measurementsA.append(detection[0])
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/02_MultipleSensorManagement.py:353]:
```suggestion
        detection = list(measurement)
        measurementsA.append(measurement.pop())
```

File: docs/tutorials/sensormanagement/02_MultipleSensorManagement.py [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+
+    chosen_targets = uncertaintymanager.choose_actions(tracksB, timestep)
+    measurementsB = []
+
+    for target in chosen_targets:
+        # The ground truth will therefore be:
+        selected_truth = truths[target][timestep]
+
+        # Observe this
+        measurement = sensor.measure([selected_truth], noise=True)
```
Comment (oharrald-Dstl) [inline docs/tutorials/sensormanagement/02_MultipleSensorManagement.py:437]: Doesn't the sensor need to change for each measurement?

**Comment by nperree-dstl** [comment]
Closing PR as tutorials will now be part of a future sensor_manager branch

**Pull request closed without merging**